go/appencryption/cache.go

```diff
@@ -96,9 +96,6 @@ func isReloadRequired(entry cacheEntry, checkInterval time.Duration) bool {
 // is not present in the cache it will retrieve the key using the provided keyLoader
 // and store the key if an error is not returned.
 func (c *keyCache) GetOrLoad(id KeyMeta, loader keyLoader) (*internal.CryptoKey, error) {
-	c.rw.Lock()
-	defer c.rw.Unlock()
-
```

Comment:
While locking at this level does incur a performance cost (as you've found) it is necessary. Moving the lock to the get/lock methods introduces a deadlock scenario.